add autolanding support

GregoireH opened the pull request:
One parameter of waypoints struct is called command. It says what this waypoint should be refering to (navigation / takeoff / land / etc)

I added a check when it reaches the waypoint, so that it would switch to autolanding behavior if the waypoint command was land, otherwise the behavior will remain the same.

jlecoeur commented:
The landing part works great, but once landed it does not continue the mission, is it normal ?
![screenshot from 2016-05-17 19-51-51](https://cloud.githubusercontent.com/assets/1379463/15333190/ff42612c-1c68-11e6-96ae-777c93a4a45d.png)
On the screenshot above, the second waypoint is a landing waypoint. After it landed and disarmed, I re-armed it, and pressed start. It rose to ~8m then went crazy to the top right corner and crashed.

GregoireH commented:
oups,

I didn't went so far, I just checked the landing / disarming procedure.

I will have a look

jlecoeur commented:
Appart from that, I am ok with how you implemented it. Would be nice to do the same with take off waypoints.

However before we handle other commands, @basil-huber we should have this meeting about how to proceed with other mission commands & waypoints.

GregoireH commented:
I don't have the crash problem, 
but I have another one while taking-off again after a landing waypoint.
The quad take-off properly and when it reaches the take-off position, it turns yaw to face north and drift towards a point that is symetric from the first take-off compare to the landing waypoint.
I guess it is related to the origin reset or something like that.

![debug_autoland2](https://cloud.githubusercontent.com/assets/3974772/15351144/05072118-1cde-11e6-8dfa-75b37e5b3a17.png)

GregoireH commented:
as you may see on the above picture,
it is only having this problem the 1st time you take off after it lands.
the second time, it took-off and remain above waypoint 1 until I pressed continue to next, then it directly went to waypoint 2...
weird

GregoireH commented:
Problem solved in 0df6a26

GregoireH commented:
It seams working fine, after merge.

2 things though:
- you set the navigation strategy to be Dubin by default => Quad's are circling around waypoints.
Is that normal ?
Should we edit the navigation_default_config.navigation_strategy in main_files in LEQuad projects ?

- if landing waypoint default altitude is 0m in newer version of QGC (v2.9.0)
Then when robots are going towards or from the landing waypoint, there will be flying at this waypoint altitude (0m) for part of the path !! => leading to collision with potential obstacles on the way...
Any idea on how to solve that ?

jlecoeur commented:
- dubin default
Yes indeed, my mistake! I changed to dubin by default because I found it was smoother than direct_to, even with very small waypoint radius. I did not think that it was circling in waypoint hold. Do you want to come back to direct_to by default ?

- altitude
In fact you can set it!

GregoireH commented:
Not ready to merge yet,
in dubin strategy the quad is circling after taking-off again on the land waypoint.

Besides, if the land waypoint is set to auto next it doesn't go to next waypoint automatically after take-off and if you press continue, it goes back to the 1st waypoint of the list

@ndousse 
Could you check please ?

GregoireH commented:
But in navigation strategy direct to 
everything is working well.

in dubin navigation
after take-off quad is circling on the landing waypoint, without adapting his yaw ?

but if we try to stop nav while it is flying it remains still. Which is what we would like to see after taking-off from land waypoint. 
Is it handled differently for a wing than for a quad ?

GregoireH commented:
Now everything should be fine.

I added a condition before going to the next,
which was on the heading or being in direct_to strategy,
and I added a or current_waypoint.command == NAV_LAND

so that if we are in NAV_LANDING, we land and if auto_next is true we set everything to point toward the next one already

Ready to merge, I think @jlecoeur @ndousse

ndousse commented:
f4b26d1 solves #255

jlecoeur commented:
@GregoireH @basil-huber I had to fix conflicts, I am not 100% sure I did it right. Can you have a look ?
- mavlink_waypoint_handler.cpp l883 l884 l891
- lequad.cpp l548 l549

GregoireH commented on communication/mavlink_waypoint_handler.cpp:
```diff
@@ -878,14 +878,17 @@ void Mavlink_waypoint_handler::auto_landing_handler()
-                state_.mav_mode_custom |= CUST_DESCENT_TO_SMALL_ALTITUDE;
+                state_.mav_mode_custom &= static_cast<Mav_mode::custom_mode_t>(0xFFFFFFE0);
+                state_.mav_mode_custom |= Mav_mode::CUST_DESCENT_TO_SMALL_ALTITUDE;
+                waypoint_hold_coordinates.waypoint = position_estimation_.local_position;
```
Nop, I think this is now done in the fonction that set the autolanding, so that you can either land on spot or at a given location.
this refer to setting waypoint_hold_coordinates

GregoireH commented on communication/mavlink_waypoint_handler.cpp:
```diff
@@ -878,14 +878,17 @@ void Mavlink_waypoint_handler::auto_landing_handler()
-                state_.mav_mode_custom |= CUST_DESCENT_TO_GND;
+                state_.mav_mode_custom &= static_cast<Mav_mode::custom_mode_t>(0xFFFFFFE0);
+                state_.mav_mode_custom |= Mav_mode::CUST_DESCENT_TO_GND;
+                waypoint_hold_coordinates.waypoint = position_estimation_.local_position;
```
But this is find !

ndousse replied:
No, I believe it's not. If you are doing Dubin navigation, you don't want to change your landing waypoint to your current location. It's better to keep always the same waypoint once you started the landing procedure

GregoireH replied:
Ok, we can remove that line, to keep landing ate the desired position

jlecoeur replied:
Done

ndousse commented on communication/mavlink_waypoint_handler.cpp:
```diff
@@ -878,14 +878,17 @@ void Mavlink_waypoint_handler::auto_landing_handler()
+                state_.mav_mode_custom &= static_cast<Mav_mode::custom_mode_t>(0xFFFFFFE0);
+                state_.mav_mode_custom |= Mav_mode::CUST_DESCENT_TO_SMALL_ALTITUDE;
+                waypoint_hold_coordinates.waypoint = position_estimation_.local_position;
+                waypoint_hold_coordinates.waypoint.pos[Z] = navigation_.takeoff_altitude/2.0f;
```
Good improvement for the drone dome! 
We could even define an close-to-ground altitude for landing. It's not obvious why you use the takeoff_altitude for the landing
